Implement `compute serve` subcommand.

Integralist opened the pull request:
Integrate a local testing environment for Compute@Edge.

Integralist commented:
@phamann this is still very much WIP but just letting you know this exists.

Integralist commented on pkg/compute/serve.go:
```diff
@@ -0,0 +1,278 @@
+	if err != nil {
+		// We presume an error executing `viceroy --version` means it isn't installed.
+		//
+		// NOTE: we can't use exec.LookPath("viceroy") because PATH is unreliable
+		// across OS platforms but also we actually install viceroy in the same
+		// location as the application configuration, which means it wouldn't be
+		// found looking up by the PATH env var.
```
@phamann The following thoughts all feel like I'm worrying about stuff that's too low-level, hence the implementation I've used doesn't try to concern itself. But that said, there are two things I'm wondering about this section:

1. I don't know if my comment "PATH is unreliable across OS platforms" is _**100%**_ accurate.

2. re: "it wouldn't be found looking up by the PATH env var" should we consider adding the config directory to the PATH (again, I'm not sure how cross platform `os.Setenv()` is**)?

> ** looking at the [`syscall.Setenv`](https://golang.org/src/syscall/env_unix.go?s=1956:1992#L86) implementation it has an OS check for Plan9 but no other OS which might not necessarily mean an issue for Windows.

Integralist commented on pkg/compute/manifest/manifest.go:
```diff
@@ -189,11 +189,20 @@ type File struct {
 	Authors         []string `toml:"authors"`
 	Language        string   `toml:"language"`
 	ServiceID       string   `toml:"service_id"`
+	Testing         Testing  `toml:"testing"`
```
5. We update the manifest `File` data structure to reflect there is now a new `[testing]` section (as per the [`fastly.toml`](https://developer.fastly.com/reference/fastly-toml/) specification).

Integralist commented on pkg/compute/manifest/manifest.go:
```diff
@@ -183,17 +183,26 @@ func (v *Version) UnmarshalText(text []byte) error {
+	Authors         []string    `toml:"authors"`
+	Language        string      `toml:"language"`
+	ServiceID       string      `toml:"service_id"`
+	LocalServer     LocalServer `toml:"local_server"`
```
5. We update the manifest `File` data structure to reflect there is now a new `[local_server]` section (as per the [`fastly.toml`](https://developer.fastly.com/reference/fastly-toml/) specification).

> **NOTE**: This doesn't reflect the current state of the world (the section is defined as `[testing]` in the design document). @triblondon and I have been pushing to rename 'testing', and after a discussion we both agreed on 'local_server' but we still need to get agreement from the wider development team.

Integralist replied:
/cc @acfoltzer @cratelyn

Integralist replied:
UPDATE: the change from `testing` to `local_server` is now agreed.

Integralist commented:
@phamann I've now marked this PR as "ready for review" although strictly speaking it's still blocked by external factors (i.e. a public repo to pull viceroy from). I'd be happy to jump on a Zoom call to run through the relevant changes (as there's lots of files that are updated, which was caused by a necessity to update the function signature for `app.Run()`, and so I can help focus in on the _actual_ changes that need reviewing).

/cc @triblondon

phamann commented on pkg/compute/serve.go:
```diff
@@ -0,0 +1,287 @@
+	c.CmdClause.Flag("name", "Package name").Action(c.name.Set).StringVar(&c.name.Value)
+	c.CmdClause.Flag("language", "Language type").Action(c.lang.Set).StringVar(&c.lang.Value)
+	c.CmdClause.Flag("include-source", "Include source code in built package").Action(c.includeSrc.Set).BoolVar(&c.includeSrc.Value)
+	c.CmdClause.Flag("force", "Skip verification steps and force build").Action(c.force.Set).BoolVar(&c.force.Value)
```
Do we also want to include a flag to skip building?

Integralist replied:
@phamann we could do, but isn't the build step (with build caching) quick enough to mean it's not a concern?

phamann replied:
I am keen to still explore adding a flag to skip building. While I agree that build+serve is the most common pattern users will want locally, however, I think in automated scenario's (i.e. CI builds) you'll most likely separate build, test, and serve (integration tests) stages and double building will slow down builds. Its also useful if you share binaries with colleagues or integrate into other scripts etc.

Integralist replied:
OK that makes sense. I'll take a look into it.

phamann commented on pkg/compute/serve.go:
```diff
@@ -0,0 +1,287 @@
+	c.CmdClause.Flag("force", "Skip verification steps and force build").Action(c.force.Set).BoolVar(&c.force.Value)
+
+	// Viceroy flags
+	c.CmdClause.Flag("env", "The environment to use when selecting backend definitions (e.g. stage)").Action(c.env.Set).StringVar(&c.env.Value)
```
While currently the only object we support is backends we will eventually allow to configure more, so consider re-wording this description. Something like:
```suggestion
	c.CmdClause.Flag("env", "The environment configuration to use (e.g. stage)").Action(c.env.Set).StringVar(&c.env.Value)
```

phamann commented on pkg/compute/serve.go:
```diff
@@ -0,0 +1,287 @@
+	segs := strings.Split(version, " ")
+
+	if len(segs) < 2 {
+		return viceroyError
```
```suggestion
		progress.Fail()
		return viceroyError
```

phamann commented on pkg/compute/serve.go:
```diff
@@ -0,0 +1,287 @@
+	installedViceroyVersion = segs[1]
+
+	if installedViceroyVersion == "" {
+		return viceroyError
```
```suggestion
		progress.Fail()
		return viceroyError
```

phamann commented on pkg/compute/serve.go:
```diff
@@ -0,0 +1,287 @@
+		env = "." + env
+	}
+	manifest := fmt.Sprintf("fastly%s.toml", env)
+	args := []string{"bin/main.wasm", "-C", manifest}
```
Not an issue now, but something for us to consider, we're making a lot of assumptions here that `fastly compute serve` was run in the root directory of a project. Is this ok? Do we want users to provide a path to teh manifest and binary via flags so it can be ran elsewhere, this may be useful in CI scenarios.

phamann commented on pkg/compute/serve.go:
```diff
@@ -0,0 +1,287 @@
+
+	cmd := common.NewStreamingExec(bin, args, os.Environ(), verbose, out)
+
+	go func(sig chan os.Signal, cmd *common.StreamingExec) {
+		<-sig
+		signal.Stop(sig)
+
+		err := cmd.Signal(os.Kill)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}(sig, cmd)
```
This seems like good default behaviour and an oversight in the original implementation of `StreamingExec`, maybe we should consider moving this logic inside `streamingexec.Exec()` so all instances of it in the CLI benefit?

phamann reviewed (changes requested):
Looking good. Couple of comments.

Integralist commented:
The macOS CI check has been failing since renaming from `master` to `main`. Tried changing the cache key entry in the workflow file (see last commit). @phamann do you have any bright ideas on why that might now be failing (no error details other than "This check was cancelled") 🤷🏻 

Otherwise I've addressed your feedback.

Integralist commented on pkg/common/exec.go:
```diff
@@ -60,3 +91,14 @@ func (s StreamingExec) Exec() error {
 }
+
+// Signal enables spawned subprocess to accept given signal.
+func (s StreamingExec) Signal(signal os.Signal) error {
```
@Integralist 

**TODO**: this doesn't need to be publicly exported anymore.

doramatadora commented on cmd/fastly/main.go:
```diff
@@ -30,13 +30,14 @@ func main() {
+		clientFactory              = app.FastlyAPIClient
+		httpClient                 = http.DefaultClient
+		cliVersioner               = update.NewGitHub(context.Background(), "fastly", "cli", "fastly")
+		viceroyVersioner           = update.NewGitHub(context.Background(), "fastly", "viceroy", "viceroy")
```
I really appreciated all these comments when reviewing, thank you 🙌🏼

Integralist commented on .github/workflows/pr_test.yml:
```diff
@@ -15,16 +15,16 @@ jobs:
       with:
         path: ~/go/bin
-        key: ${{ runner.os }}-go-bin-${{ hashFiles('**/go.sum') }}
+        key: ${{ runner.os }}-gobin-${{ hashFiles('**/go.sum') }}
```
I think I can revert these 👍🏻

Integralist commented:
We might need to watch out for errors going to stdout rather than stderr https://github.com/fastly/Viceroy/pull/171#discussion_r656566379

phamann commented on pkg/compute/serve.go:
```diff
@@ -0,0 +1,309 @@
+		return err
+	}
+
+	manifest := filepath.Join(wd, fmt.Sprintf("fastly%s.toml", env))
```
Am I interpreting this correctly that a `fastly compute serve --env test` will require a file named `fastlytest.toml` noting the lack of any separator between fastly and the stage? I think it would be a nicer ux if we add one:

```suggestion
	manifest := filepath.Join(wd, fmt.Sprintf("fastly-%s.toml", env))
```

Integralist replied:
@phamann there's a `.` inbetween so `fastly.test.toml` look a couple lines up from that one I think.

Integralist replied:
Lines 277-279

phamann replied:
Apologies, my fault, thats perfect 👌🏻

phamann commented on pkg/compute/serve.go:
```diff
@@ -0,0 +1,309 @@
+	c.CmdClause.Flag("include-source", "Include source code in built package").Action(c.includeSrc.Set).BoolVar(&c.includeSrc.Value)
+	c.CmdClause.Flag("force", "Skip verification steps and force build").Action(c.force.Set).BoolVar(&c.force.Value)
+
+	// Viceroy flags
```
To complement my request for a skip building flag above, I think it would be nice if we also allowed users to provide the binary file to run as an optional flag `--file` or `--bin`. This could default to `bin/main.wasm`.

```suggestion
	// Viceroy flags
        c.CmdClause.Flag("file", "The Wasm file to run").Action(c.file.Set).StringVar(&c.file.Value)	
```

Integralist replied:
OK. I'll take a look.

phamann commented on pkg/compute/serve.go:
```diff
@@ -0,0 +1,309 @@
+	}
+
+	manifest := filepath.Join(wd, fmt.Sprintf("fastly%s.toml", env))
+	args := []string{"bin/main.wasm", "-C", manifest}
```
Per my comment above in this file, I don't like how we have hard coded `bin/main.wasm` here and would prefer if this was configurable.

Integralist replied:
👍🏻

phamann commented on pkg/update/check.go:
```diff
@@ -24,6 +25,9 @@ func Check(ctx context.Context, currentVersion string, cliVersioner Versioner) (
 		return current, latest, false, fmt.Errorf("error fetching latest version: %w", err)
 	}
 
+	asset := fmt.Sprintf("%s_v%s_%s-%s.tar.gz", cliVersioner.Binary(), latest, runtime.GOOS, runtime.GOARCH)
```
This template string is used in a couple of places across the code base, maybe consider moving it to a package level constant.

Integralist replied:
👍🏻

phamann reviewed (changes requested):
Looking good, a couple of requests around exposing more flags for a nicer UX.

phamann commented on pkg/compute/serve.go:
```diff
@@ -0,0 +1,313 @@
+	c.manifest.File.Read(manifest.Filename)
+
+	c.CmdClause.Flag("env", "The environment configuration to use (e.g. stage)").Action(c.env.Set).StringVar(&c.env.Value)
+	c.CmdClause.Flag("file", "The Wasm file to run").StringVar(&c.file)
```
Am interested in knowing why you didn't set the default via the flag itself. I personally prefer this method, as the default value then gets documented rather than being magic. 

```suggestion
	c.CmdClause.Flag("file", "The Wasm file to run").Default("bin/main.wasm").StringVar(&c.file)
```

Integralist replied:
I didn't know that was a thing with Kingpin. So I'll sort that 👍🏻

phamann reviewed (approved):
👍🏻  Looks good to me, other than minor non-blocking comment.

dkegel-fastly commented:
An --addr option would let me use this instead of directly running Viceroy.
I can't push a branch, so here's a git patch that implements one:
[0001-serve-accept-optional-addr-flag-and-pass-along-to-Vi.patch.txt](https://github.com/fastly/cli/files/6812622/0001-serve-accept-optional-addr-flag-and-pass-along-to-Vi.patch.txt)

kailan reviewed (approved):
All seems to work as intended 👍 🚀
